flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,734 @@
+            }
+        }
+
+        public static Map<Integer, Integer> getActualResult() {
+            return sourceList.stream()
+                    .collect(
+                            Collectors.toConcurrentMap(
+                                    Function.identity(), element -> 1, Integer::sum));
+        }
+
+        @Override
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
+            lastSentStored.update(Arrays.asList(currentIndex, completedCheckpointNum.get()));
+        }
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {
+            lastSentStored =
+                    context.getOperatorStateStore()
+                            .getListState(new ListStateDescriptor<>("counter", Integer.class));
+            if (context.isRestored()) {
+                Iterable<Integer> indexAndCheckpointNum = lastSentStored.get();
+                currentIndex = get(indexAndCheckpointNum, 0);
+                completedCheckpointNum.compareAndSet(0, get(indexAndCheckpointNum, 1));
+            }
+        }
+
+        @Override
+        public void run(SourceContext<Integer> ctx) throws Exception {
+            while (!isCanceled && currentIndex < sourceList.size()) {
+                preHandle();
+                synchronized (ctx.getCheckpointLock()) {
+                    if (!isCanceled && currentIndex < sourceList.size()) {
+                        int currentElement = sourceList.get(currentIndex++);
+                        ctx.collect(currentElement);
+                    }
+                }
+            }
+        }
+
+        protected void preHandle() throws Exception {
+            // do nothing by default
+        }
+
+        protected void waitUntilFalse(SerializableBooleanSupplier supplier)
+                throws InterruptedException {
+            while (supplier.getAsBoolean()) {
+                Thread.sleep(10);
+            }
+        }
+
+        protected void throwArtificialFailure() throws Exception {
+            throw new Exception("Artificial Failure");
+        }
+
+        @Override
+        public void cancel() {
+            isCanceled = true;
+        }
+
+        @Override
+        public void notifyCheckpointComplete(long checkpointId) throws Exception {
+            completedCheckpointNum.getAndIncrement();
+        }
+    }
+
+    /** A sink puts the result of WordCount into a Map using for validating. */
+    private static class CollectionSink implements SinkFunction<Tuple2<Integer, Integer>> {
+
+        private static final Map<Integer, Integer> expectedResult = new ConcurrentHashMap<>();
+
+        @Override
+        public synchronized void invoke(Tuple2<Integer, Integer> value) throws Exception {
+            expectedResult.put(
+                    value.f0, Math.max(value.f1, expectedResult.getOrDefault(value.f0, 0)));
```

Comment:
Could you explain the use of `Math.max` here? Isn't it potentially hiding state inconsistency (i.e. when the upstream updated its state with a count, but that update was lost)?

Reply:
It's just used to mock simple wordcount example (To maintain the wordcount result from `CountMapper`).
I think the result will be updated synchronously because the `CountMapper` will always pass the newest state to the sink and the parallism of the sink is always 1.

Reply:
Let's consider the following example with the key `7`:
1. mapper sends `(7 -> 1)`, sink collects `(7 -> 1)`
2. checkpoint starts and completes: mapper saves  `(7 -> 1)` in its state
3. mapper sends `(7 -> 2)`, sink collects `(7 -> 2)`
4. failure, recovery, mapper restores to the **wrong** state `(7 -> 0)` (should be 1)
5. mapper sends `(7 -> 1)`, but sink chooses `(7 -> 2)` - because of `max`

So the wrong state in (4) will not be detected.
Or am I missing something?

Reply:
Why in the step 4, the mapper will restores to the  `(7 -> 0)` not `(7 -> 1)` which has been  in the newest checkpoint ? 
I think the result will be eventually consistent, not strictly consistent. right ?
Or maybe I missed something ?

Reply:
I mean if there is such a bug in backend, for example it restores to an empty state; then the test wouldn't catch it.